Support HTTP POST method for /search and /count

PJGitLan opened the pull request:
Additional POST method accepting the search filter as payload. Workaround for longer RQL filters until QUERY is possible to implement.

thjaeckle commented on gateway/service/src/main/java/org/eclipse/ditto/gateway/service/endpoints/routes/thingsearch/ThingSearchRoute.java:
```diff
@@ -79,10 +79,28 @@ public Route buildSearchRoute(final RequestContext ctx, final DittoHeaders ditto
-                CountThings.of(calculateFilter(params.get(ThingSearchParameter.FILTER)),
-                        calculateNamespaces(params.get(ThingSearchParameter.NAMESPACES)),
-                        dittoHeaders))));
+        // POST things/count?namespaces=<namespacesString>
```
Please move that comment to where the `post()` is called

thjaeckle commented on gateway/service/src/main/java/org/eclipse/ditto/gateway/service/endpoints/routes/thingsearch/ThingSearchRoute.java:
```diff
@@ -79,10 +79,28 @@ public Route buildSearchRoute(final RequestContext ctx, final DittoHeaders ditto
+                            calculateNamespaces(params.get(ThingSearchParameter.NAMESPACES)),
+                                    dittoHeaders)))),
+                post(() -> thingSearchParameterOptional(
+                        params -> ensureMediaTypeJsonWithFallbacksThenExtractDataBytes(
```
By using `ensureMediaTypeJsonWithFallbacksThenExtractDataBytes` you enforce the POST payload to be of `application/json` - do you really want that?
Did you test that?

To me, this looks like it could not be what is wanted - as the `filter` below inserted into `CountThings` is certainly not a Json, but a plain text string.

I would expect the content-type to be `application/x-www-form-urlencoded` instead  and that e.g. also query params would be provided all as form values - instead of mixing those.

thjaeckle commented on gateway/service/src/main/java/org/eclipse/ditto/gateway/service/endpoints/routes/thingsearch/ThingSearchRoute.java:
```diff
@@ -96,13 +114,43 @@ private Route searchThings(final RequestContext ctx, final DittoHeaders dittoHea
-                                .orElse(null),
-                        calculateNamespaces(params.get(ThingSearchParameter.NAMESPACES)),
-                        dittoHeaders))));
+        // POST things?options=<optionsString>PUT
```
`POST things?options=<optionsString>PUT` ? 
What is that supposed to mean?

As mentioned above, I would not "mix" query params with form params of the POST - and provide all query params as form values instead.

PJGitLan replied:
Typo

thjaeckle commented on gateway/service/src/test/java/org/eclipse/ditto/gateway/service/endpoints/routes/thingsearch/ThingSearchRouteTest.java:
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2017 Contributors to the Eclipse Foundation
```
Please adjust creation year of the file to 2023

thjaeckle commented on gateway/service/src/test/java/org/eclipse/ditto/gateway/service/endpoints/routes/thingsearch/ThingSearchRouteTest.java:
```diff
@@ -0,0 +1,75 @@
+    public void countThingsShouldGetFilterFromBody() {
+        final var filter = "and(like(definition,\"*test*\"))";
+        final var result = underTest.run(HttpRequest.POST("/search/things/count")
+                .withEntity(ContentTypes.APPLICATION_JSON, filter));
```
Please also adjust to `application/x-www-form-urlencoded` content-type.

PJGitLan replied:
done

thjaeckle reviewed (changes requested):
Thanks for the PR.

Please adjust the `POST` search to take all parameters from form parameters and adjust the supported content-type for that to `application/x-www-form-urlencoded` (form encoding).

AFAIK that is the best practice do do "POST" as replacement for "GET" with too long query parameters.

We would then also need to update some documentation (OpenAPI and also Ditto docs) as part of the PR - but one step after another :)

thjaeckle commented on gateway/service/src/main/java/org/eclipse/ditto/gateway/service/endpoints/routes/thingsearch/ThingSearchRoute.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017 Contributors to the Eclipse Foundation
```
This should stay at 2017 - the file creation year.
I referenced the other file, the unit test which was added.

PJGitLan replied:
oops 😅

thjaeckle commented:
@PJGitLan I had a brief look at the implementation - and it looks quite good. 👍 
I will invest more time in manually testing ..

Now we would only need to provide documentation for this in order to get it merged.
As mentioned before:
> We would then also need to update some documentation (OpenAPI and also Ditto docs) as part of the PR - but one step after another :)

thjaeckle commented:
@PJGitLan wanted to ping you - are you still interested in getting this PR merged?

PJGitLan commented:
> @PJGitLan wanted to ping you - are you still interested in getting this PR merged?

yes, I am still want to get this merged.  Sorry for the long wait.

thjaeckle commented:
Status: Waiting for documentation - as stated before.

thjaeckle commented:
@PJGitLan thanks a lot - I will review soon.
Should then hopefully make it to the upcoming Ditto 3.4.0 :)

thjaeckle commented:
@PJGitLan I added a commit to your PR and hope you don't mind
I noticed that the form parameter parsing was handmade and quite complex + error prone, so I reworked that part to use Akka HTTP's `formFieldMultiMap()` API to extract all form parameters in a very similar way to how we extract the query parameters.

This implementation is now also way more tolerant how the `filter` and so on parameters are provided - as far as I see, they don't need to be URL encoded which makes it a lot easier.

Please check my changes and if you can live with them ;)

PJGitLan commented:
> @PJGitLan I added a commit to your PR and hope you don't mind I noticed that the form parameter parsing was handmade and quite complex + error prone, so I reworked that part to use Akka HTTP's `formFieldMultiMap()` API to extract all form parameters in a very similar way to how we extract the query parameters.
> 
> This implementation is now also way more tolerant how the `filter` and so on parameters are provided - as far as I see, they don't need to be URL encoded which makes it a lot easier.
> 
> Please check my changes and if you can live with them ;)

Changes look good to me. I didn't know there was a 'formFieldMultiMap()' method. 
Thanks for the help.
